### PR TITLE
Use a DiscoverProvider-function to discover the bootstrap hosts

### PR DESCRIPTION
**Pull request by mennopruijssers**
By introducing the concept of DiscoverProviders in ringpop-node, we make it possible in the future to reload the host list. This is a requirement for partition healing.

When the number of hosts or their ip/ports are more dynamic (e.g. when bootstrapping over hyperbahn) it's going to be useful to reload the hosts after bootstrapping.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -500,45 +517,24 @@ RingPop.prototype.lookupN = function lookupN(key, n) {
 };
 
 RingPop.prototype.reload = function reload(file, callback) {
```
Comment (mennopruijssers) [inline index.js]: I'm not sure if/how we want to support this function. It's currently called from the `admin/reload`-endpoint but only supports a file.

File: discover-providers.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        return callback(null, hosts);
+    };
+}
+
+/**
+ * Creates a DiscoverProvider from the provided options.
+ *
+ * @param opts: an object configuring the DiscoverProvider:
+ *          - if opts or opts.discoverProvider is a function, it's returned as the {DiscoverProvider}.
+ *          - if opts.hosts is an array, a static hosts discover provider is returned (see {createStaticHostsProvider})
```
Comment (thanodnl) [inline discover-providers.js]: if opts or opts.hosts ....

File: discover-providers.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ *          - if opts.bootstrapFile is a string, a json file discover provider is returned (see {createJsonFileDiscoverProvider})
```
Comment (thanodnl) [inline discover-providers.js]: if opts or opts.bootstrapFile ....

File: index.js (hunk 1 above)
Comment (thanodnl) [inline index.js]: I think the right thing to do here is to add a `// Deprecated` comment to the `/admin/reload` handler and add a warning log when that endpoint is called. Later we can remove it.

**Comment by thanodnl** [comment]
LGTM

File: discover-providers.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    return null;
+}
+
+module.exports = {
+    createFromOpts: createFromOpts,
+    createStaticHostsProvider: createStaticHostsProvider,
+    createJsonFileDiscoverProvider: createJsonFileDiscoverProvider
+};
```
Comment (dansimau) [inline discover-providers.js]: Missing EOL

File: index.js (hunk 1 above)
Comment (dansimau) [inline index.js]: Agree with @thanodnl. I'd also check with @jwolski that it's OK to deprecate. I.e. how confident are we this is not in use?

File: test/unit/discover-providers-example.json [hunk 5]
```diff
@@ -0,0 +1 @@
+["127.0.0.1:3000", "127.0.0.1:3001"]
```
Comment (dansimau) [inline test/unit/discover-providers-example.json]: Missing EOL

File: index.js (hunk 1 above)
Comment (mennopruijssers) [inline index.js]: It's not used in ringpop-admin and (according to our stats backend) never called in production.

**Comment by mennopruijssers** [comment]
Based on feedback from @jwolski, I moved the deprecation of the endpoint to #255

**Comment by mennopruijssers** [comment]
@dansimau All your (style) feedback is taking care of! Let's try to come up with a better linter ;).

**Comment by dansimau** [comment]
LGTM 👍

**Pull request merged**